unstoppable-ios-app/domains-manager-ios/Entities/Storages/DefaultsStorage.swift

```diff
@@ -13,34 +13,34 @@ class DefaultsStorage<T: Equatable> where T: Codable {
     }
     var storageKey: String = ""
     var q = DispatchQueue(label: "work-queue")
-
-    func retrieveAll() -> [T] {
-        guard let arrayObject = UserDefaults.standard
-            .object(forKey: storageKey) as? Data else { return [] }
-        guard let array = try? JSONDecoder().decode([T].self, from: arrayObject) else {
-            Debugger.printFailure("Failed to decode connection intents from data: \(String(data: arrayObject, encoding: .utf8) ?? "n/a")", critical: false)
-            return []
-        }
-        return array
-    }
         
-    func store(elements: [T]) throws {
-        let appsData = try JSONEncoder().encode(elements)
-        UserDefaults.standard.set(appsData, forKey: storageKey)
-    }
-    
     func save(newElement: T) {
         q.async {
-            var elements = self.retrieveAll()
+            var elements = self.getAllFromDefaults()
             elements.update(with: newElement)
             try? self.store(elements: elements)
         }
-    }    
+    }
+    
+    func retrieveAll() -> [T] {
+        var result: [T] = []
+        if Thread.isMainThread {
+            DispatchQueue.global().sync {
+                q.sync {
+                    result = getAllFromDefaults()
+                }
+            }
+        } else {
```

Comment:
What is the logic behind checking for main thread and blocking it with additional global queue?

Reply:
If it is the main thread, the blocking code is thrown into _another_ thread. Otherwise we are free to block

Reply:
`q.sync {}` still blocks the main thread, but combined with `DispatchQueue.global().sync` it does not